### PR TITLE
bmt pkg migrate under swarm

### PR DESCRIPTION
**Pull request by zelig**
This PR is putting back bmt under swarm. This package implements a crucial component of swarm which in turn is only relevant for swarm, hence its natural place is under bee.

The PR does not only migrate the existing repo in https://github.com/ethersphere/bmt to `pkg/bmt`, but also
- simplify and clean up code
- remove unused/useless function
- reorganise tree and hasher fields
- get rid of locking completely (was originally, but got corrupted on the way).. also tested with `-race`
- get rid of section on hasher, use and reuse buffer on tree

### CODE REVIEW

**Comment by zelig** [comment]
OLD:

```

goos: darwin
goarch: amd64
pkg: github.com/ethersphere/bee/pkg/bmt/legacy
BenchmarkBMT
BenchmarkBMT/SHA3_size_4096
BenchmarkBMT/SHA3_size_4096-16         	  110826	     10365 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_4096
BenchmarkBMT/Baseline_size_4096-16     	   42672	     28132 ns/op	   68625 B/op	     391 allocs/op
BenchmarkBMT/REF_size_4096
BenchmarkBMT/REF_size_4096-16          	   14868	     79594 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_4096
BenchmarkBMT/BMT_size_4096-16          	   25887	     46000 ns/op	   70093 B/op	     450 allocs/op
BenchmarkBMT/SHA3_size_2048
BenchmarkBMT/SHA3_size_2048-16         	  207780	      5558 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_2048
BenchmarkBMT/Baseline_size_2048-16     	   66187	     17890 ns/op	   35856 B/op	     199 allocs/op
BenchmarkBMT/REF_size_2048
BenchmarkBMT/REF_size_2048-16          	   14996	     79708 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_2048
BenchmarkBMT/BMT_size_2048-16          	   38167	     31216 ns/op	   35788 B/op	     229 allocs/op
BenchmarkBMT/SHA3_size_1024
BenchmarkBMT/SHA3_size_1024-16         	  356989	      2932 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_1024
BenchmarkBMT/Baseline_size_1024-16     	   94888	     12043 ns/op	   19472 B/op	     103 allocs/op
BenchmarkBMT/REF_size_1024
BenchmarkBMT/REF_size_1024-16          	   14995	     80365 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_1024
BenchmarkBMT/BMT_size_1024-16          	   51885	     22686 ns/op	   18890 B/op	     120 allocs/op
BenchmarkBMT/SHA3_size_512
BenchmarkBMT/SHA3_size_512-16          	  720728	      1607 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_512
BenchmarkBMT/Baseline_size_512-16      	  127203	      8799 ns/op	   11280 B/op	      55 allocs/op
BenchmarkBMT/REF_size_512
BenchmarkBMT/REF_size_512-16           	   15009	     79165 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_512
BenchmarkBMT/BMT_size_512-16           	   68422	     17003 ns/op	   10698 B/op	      67 allocs/op
BenchmarkBMT/SHA3_size_256
BenchmarkBMT/SHA3_size_256-16          	 1305750	       910 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_256
BenchmarkBMT/Baseline_size_256-16      	  181760	      6308 ns/op	    7184 B/op	      31 allocs/op
BenchmarkBMT/REF_size_256
BenchmarkBMT/REF_size_256-16           	   14961	     79719 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_256
BenchmarkBMT/BMT_size_256-16           	   94028	     11856 ns/op	    6857 B/op	      42 allocs/op
BenchmarkBMT/SHA3_size_128
BenchmarkBMT/SHA3_size_128-16          	 2114281	       561 ns/op	     512 B/op	       3 allocs/op
BenchmarkBMT/Baseline_size_128
BenchmarkBMT/Baseline_size_128-16      	  240338	      4807 ns/op	    5136 B/op	      19 allocs/op
BenchmarkBMT/REF_size_128
BenchmarkBMT/REF_size_128-16           	   14779	     79851 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_128
BenchmarkBMT/BMT_size_128-16           	  126242	      9077 ns/op	    5193 B/op	      31 allocs/op
BenchmarkPool
BenchmarkPool/poolsize_1_size_4096
BenchmarkPool/poolsize_1_size_4096-16  	     243	   4908052 ns/op	 7014276 B/op	   45105 allocs/op
BenchmarkPool/poolsize_8_size_4096
BenchmarkPool/poolsize_8_size_4096-16  	     388	   3470043 ns/op	 7015787 B/op	   45116 allocs/op
BenchmarkPool/poolsize_1_size_2048
BenchmarkPool/poolsize_1_size_2048-16  	     358	   3445971 ns/op	 3583870 B/op	   23005 allocs/op
BenchmarkPool/poolsize_8_size_2048
BenchmarkPool/poolsize_8_size_2048-16  	     733	   1742984 ns/op	 3584601 B/op	   23011 allocs/op
BenchmarkPool/poolsize_1_size_1024
BenchmarkPool/poolsize_1_size_1024-16  	     476	   2486565 ns/op	 1894081 B/op	   12104 allocs/op
BenchmarkPool/poolsize_8_size_1024
BenchmarkPool/poolsize_8_size_1024-16  	    1383	    925249 ns/op	 1894497 B/op	   12107 allocs/op
BenchmarkPool/poolsize_1_size_512
BenchmarkPool/poolsize_1_size_512-16   	     607	   1942656 ns/op	 1074741 B/op	    6803 allocs/op
BenchmarkPool/poolsize_8_size_512
BenchmarkPool/poolsize_8_size_512-16   	    2181	    567483 ns/op	 1074975 B/op	    6804 allocs/op
BenchmarkPool/poolsize_1_size_256
BenchmarkPool/poolsize_1_size_256-16   	     787	   1493094 ns/op	  690706 B/op	    4302 allocs/op
BenchmarkPool/poolsize_8_size_256
BenchmarkPool/poolsize_8_size_256-16   	    2997	    406166 ns/op	  690837 B/op	    4303 allocs/op
BenchmarkPool/poolsize_1_size_128
BenchmarkPool/poolsize_1_size_128-16   	     976	   1174681 ns/op	  524451 B/op	    3204 allocs/op
BenchmarkPool/poolsize_8_size_128
BenchmarkPool/poolsize_8_size_128-16   	    3717	    328042 ns/op	  524346 B/op	    3202 allocs/op
PASS
ok  	github.com/ethersphere/bee/pkg/bmt/legacy	55.028s
```


NEW:

```

 :) go test -v ./... -run none -bench .
goos: darwin
goarch: amd64
pkg: github.com/ethersphere/bee/pkg/bmt
BenchmarkBMT
BenchmarkBMT/SHA3_size_4096
BenchmarkBMT/SHA3_size_4096-16         	  102841	     10726 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_4096
BenchmarkBMT/Baseline_size_4096-16     	   33118	     36761 ns/op	  126003 B/op	     637 allocs/op
BenchmarkBMT/REF_size_4096
BenchmarkBMT/REF_size_4096-16          	   14910	     79523 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_4096
BenchmarkBMT/BMT_size_4096-16          	   23839	     49768 ns/op	   66048 B/op	     386 allocs/op
BenchmarkBMT/SHA3_size_2048
BenchmarkBMT/SHA3_size_2048-16         	  197842	      5816 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_2048
BenchmarkBMT/Baseline_size_2048-16     	   53834	     21781 ns/op	   62513 B/op	     317 allocs/op
BenchmarkBMT/REF_size_2048
BenchmarkBMT/REF_size_2048-16          	   14665	     80115 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_2048
BenchmarkBMT/BMT_size_2048-16          	   35178	     34312 ns/op	   36864 B/op	     215 allocs/op
BenchmarkBMT/SHA3_size_1024
BenchmarkBMT/SHA3_size_1024-16         	  375714	      3108 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_1024
BenchmarkBMT/Baseline_size_1024-16     	   82172	     14215 ns/op	   30768 B/op	     157 allocs/op
BenchmarkBMT/REF_size_1024
BenchmarkBMT/REF_size_1024-16          	   14696	     80702 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_1024
BenchmarkBMT/BMT_size_1024-16          	   46714	     25244 ns/op	   20480 B/op	     119 allocs/op
BenchmarkBMT/SHA3_size_512
BenchmarkBMT/SHA3_size_512-16          	  680788	      1767 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_512
BenchmarkBMT/Baseline_size_512-16      	  117007	     10059 ns/op	   14896 B/op	      77 allocs/op
BenchmarkBMT/REF_size_512
BenchmarkBMT/REF_size_512-16           	   14766	     80693 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_512
BenchmarkBMT/BMT_size_512-16           	   60512	     19450 ns/op	   12288 B/op	      71 allocs/op
BenchmarkBMT/SHA3_size_256
BenchmarkBMT/SHA3_size_256-16          	 1000000	      1084 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_256
BenchmarkBMT/Baseline_size_256-16      	  151408	      7523 ns/op	    6960 B/op	      37 allocs/op
BenchmarkBMT/REF_size_256
BenchmarkBMT/REF_size_256-16           	   14575	     81466 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_256
BenchmarkBMT/BMT_size_256-16           	   78123	     14861 ns/op	    8192 B/op	      47 allocs/op
BenchmarkBMT/SHA3_size_128
BenchmarkBMT/SHA3_size_128-16          	 1642686	       718 ns/op	     992 B/op	       5 allocs/op
BenchmarkBMT/Baseline_size_128
BenchmarkBMT/Baseline_size_128-16      	  208344	      5579 ns/op	    2992 B/op	      17 allocs/op
BenchmarkBMT/REF_size_128
BenchmarkBMT/REF_size_128-16           	   14575	     81770 ns/op	   73152 B/op	     445 allocs/op
BenchmarkBMT/BMT_size_128
BenchmarkBMT/BMT_size_128-16           	   94507	     12403 ns/op	    6144 B/op	      35 allocs/op
BenchmarkPool
BenchmarkPool/poolsize_1_size_4096
BenchmarkPool/poolsize_1_size_4096-16  	     244	   4867025 ns/op	 6624285 B/op	   38910 allocs/op
BenchmarkPool/poolsize_16_size_4096
BenchmarkPool/poolsize_16_size_4096-16 	     340	   3754898 ns/op	 6624724 B/op	   38913 allocs/op
BenchmarkPool/poolsize_1_size_2048
BenchmarkPool/poolsize_1_size_2048-16  	     342	   3508533 ns/op	 3705500 B/op	   21806 allocs/op
BenchmarkPool/poolsize_16_size_2048
BenchmarkPool/poolsize_16_size_2048-16 	     607	   2041333 ns/op	 3705795 B/op	   21809 allocs/op
BenchmarkPool/poolsize_1_size_1024
BenchmarkPool/poolsize_1_size_1024-16  	     446	   2688582 ns/op	 2067042 B/op	   12206 allocs/op
BenchmarkPool/poolsize_16_size_1024
BenchmarkPool/poolsize_16_size_1024-16 	    1077	   1039102 ns/op	 2067278 B/op	   12208 allocs/op
BenchmarkPool/poolsize_1_size_512
BenchmarkPool/poolsize_1_size_512-16   	     574	   2072540 ns/op	 1247668 B/op	    7404 allocs/op
BenchmarkPool/poolsize_16_size_512
BenchmarkPool/poolsize_16_size_512-16  	    1700	    638601 ns/op	 1247763 B/op	    7405 allocs/op
BenchmarkPool/poolsize_1_size_256
BenchmarkPool/poolsize_1_size_256-16   	     670	   1761121 ns/op	  837976 B/op	    5003 allocs/op
BenchmarkPool/poolsize_16_size_256
BenchmarkPool/poolsize_16_size_256-16  	    2629	    439419 ns/op	  838058 B/op	    5004 allocs/op
BenchmarkPool/poolsize_1_size_128
BenchmarkPool/poolsize_1_size_128-16   	     799	   1479237 ns/op	  633160 B/op	    3803 allocs/op
BenchmarkPool/poolsize_16_size_128
BenchmarkPool/poolsize_16_size_128-16  	    3346	    355854 ns/op	  633261 B/op	    3804 allocs/op
PASS
ok  	github.com/ethersphere/bee/pkg/bmt	53.677s
```

File: pkg/bmt/bmt.go [hunk 1]
```diff
@@ -1,33 +1,255 @@
...
+	span   []byte      // The span of the data subsumed under the chunk
+}
+
+// Capacity returns the maximum amount of bytes that will be processed by this hasher implementation.
+func (h *Hasher) Capacity() int {
+	return h.maxSize
+}
+
+// writeSection allows asynchronous writes of the base segments
+func (h *Hasher) WriteSection(idx int, data []byte) {
```
Comment (acud) [inline pkg/bmt/bmt.go:43]: can we please remove this stuff, also from the interface

File: pkg/bmt/bmt.go [hunk 2]
```diff
@@ -1,33 +1,255 @@
...
+	} else {
+		h.writeNode(n, isLeft, section)
+	}
+}
 
-	// SetSpanBytes sets the length prefix of BMT hash in byte form.
-	SetSpanBytes([]byte) error
+// calculates the hash of the data using hash.Hash.
+//
+// BUG: This legacy implementation has no error handling for the writer. Use with caution.
```
Comment (acud) [inline pkg/bmt/bmt.go:158]: maybe would be good to actually address this as part of the cleanup

File: pkg/bmt/doc.go [hunk 3]
```diff
@@ -1,6 +1,34 @@
-// Copyright 2020 The Swarm Authors. All rights reserved.
+// Copyright 2021 The Swarm Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// Package bmt defines the interface for the Binary Merkle Tree hash.
+// Binary Merkle Tree Hash is a hash function over arbitrary datachunks of limited size.
```
Comment (acud) [inline pkg/bmt/doc.go:5]: `// Package bmt....`

File: pkg/bmt/interface.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	SetSpan(int64)
+
+	// SetSpanBytes sets the length prefix of BMT hash in byte form.
+	SetSpanBytes([]byte)
+
+	// Capacity returns the maximum amount of bytes that will be processed by the implementation.
+	Capacity() int
+
+	// WriteSection writes to a specific section of the data to be hashed.
+	WriteSection(idx int, data []byte)
```
Comment (acud) [inline pkg/bmt/interface.go:32]: let's remove this dead code plz

File: pkg/bmt/interface.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+// Hash provides the necessary extension of the hash interface to add the length-prefix of the BMT hash.
+//
+// Any implementation should make it possible to generate a BMT hash using the hash.Hash interface only.
+// However, the limitation will be that the Span of the BMT hash always must be limited to the amount of bytes actually written.
+type Hash interface {
+	hash.Hash
+
+	// SetSpan sets the length prefix of BMT hash.
+	SetSpan(int64)
```
Comment (acud) [inline pkg/bmt/interface.go:23]: we never use `SetSpan`, and I would be happy to get rid of it too.

File: pkg/bmt/interface.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+	// SetSpanBytes sets the length prefix of BMT hash in byte form.
+	SetSpanBytes([]byte)
+
+	// Capacity returns the maximum amount of bytes that will be processed by the implementation.
```
Comment (acud) [inline pkg/bmt/interface.go:28]: this comment is pretty obscure

File: pkg/bmt/reference/reference_test.go [hunk 7]
```diff
@@ -44,52 +45,56 @@ func sha3hash(t *testing.T, data ...[]byte) []byte {
 func TestRefHasher(t *testing.T) {
 	// the test struct is used to specify the expected BMT hash for
 	// segment counts between from and to and lengths from 1 to datalength
-	for i, x := range []struct {
+	type test struct {
```
Comment (acud) [inline pkg/bmt/reference/reference_test.go:48]: why revert this change? i inlined them all together so that we have a nice table driven test

File: pkg/bmt/reference/reference_test.go [hunk 8]
```diff
@@ -98,12 +103,15 @@ func TestRefHasher(t *testing.T) {
...
+
+	// run the tests
+	for _, x := range tests {
 		for segCount := x.from; segCount <= x.to; segCount++ {
 			for length := 1; length <= segCount*32; length++ {
 				t.Run(fmt.Sprintf("%d_segments_%d_bytes", segCount, length), func(t *testing.T) {
-					g := mockbytes.New(i, mockbytes.MockTypeStandard)
-					data, err := g.RandomBytes(length)
+					data := make([]byte, length)
+					_, err := io.ReadFull(crand.Reader, data)
```
Comment (acud) [inline pkg/bmt/reference/reference_test.go:114]: be aware that `mockbytes` generate random bytes with a reproducible seed. here you are switching to random tests which we cannot debug since we don't have the seed anymore

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/bmt/interface.go [hunk 9]
```diff
@@ -19,15 +19,15 @@ const (
...
 	hash.Hash
 
-	// SetSpan sets the length prefix of BMT hash.
-	SetSpan(int64)
+	// SetMetaLength sets the length prefix of BMT hash.
+	SetMetaToLength(int64)
 
 	// SetSpanBytes sets the length prefix of BMT hash in byte form.
-	SetSpanBytes([]byte)
+	SetMetaBytes([]byte)
```
Comment (acud) [inline pkg/bmt/interface.go:26]: amend comment

File: pkg/bmt/interface.go [hunk 10]
```diff
@@ -19,15 +19,15 @@ const (
 type Hash interface {
 	hash.Hash
 
-	// SetSpan sets the length prefix of BMT hash.
-	SetSpan(int64)
+	// SetMetaLength sets the length prefix of BMT hash.
+	SetMetaToLength(int64)
```
Comment (acud) [inline pkg/bmt/interface.go:23]: comment says `SetMetaLength` but method name is `SetMetaToLength`. Also, I would even go as far as renaming it to be `SetMetaInt64`, since the number can be literally anything. And a small naming nitpick - consider `Preamble` or `Header` instead of `Meta`

File: pkg/file/pipeline/builder/builder_test.go [hunk 11]
```diff
@@ -71,7 +71,7 @@ func TestEmpty(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	exp := swarm.MustParseHexAddress("ffd70157e48063fc33c97a050f7f640233bf646cc98d9524c6b92bcf3ab56f83")
+	exp := swarm.MustParseHexAddress("b34ca8c22b9e982354f9c7f50b470d66db428d880c8a904d5fe4ec9713171526")
```
Comment (acud) [inline pkg/file/pipeline/builder/builder_test.go:74]: whoops... how is this possible that the zero hash changed now?

File: pkg/file/pipeline/bmt/bmt_test.go [hunk 12]
```diff
@@ -29,9 +28,9 @@ func TestBmtWriter(t *testing.T) {
 		{
 			// this is a special case, since semantically it can be considered the hash
 			// of an empty file (since data is all zeros).
-			name:    "all zeros",
-			data:    make([]byte, swarm.ChunkSize),
-			expHash: mustDecodeString(t, "09ae927d0f3aaa37324df178928d3826820f3dd3388ce4aaebfc3af410bde23a"),
+			name:    "empty file",
+			data:    make([]byte, 0),
+			expHash: mustDecodeString(t, "b34ca8c22b9e982354f9c7f50b470d66db428d880c8a904d5fe4ec9713171526"),
```
Comment (acud) [inline pkg/file/pipeline/bmt/bmt_test.go:33]: this hash is different than the current zero hash which is strange. also, it might be worth to change the comment of the test, which is now wrong

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/file/pipeline/builder/builder_test.go (hunk 11 above)
Comment (zelig) [reply pkg/file/pipeline/builder/builder_test.go:74]: it was incorrect what can i do :) ?

File: pkg/file/pipeline/bmt/bmt_test.go (hunk 12 above)
Comment (zelig) [reply pkg/file/pipeline/bmt/bmt_test.go:33]: it was incorrect what can i do :) ?

File: pkg/file/pipeline/builder/builder_test.go (hunk 11 above)
Comment (acud) [reply pkg/file/pipeline/builder/builder_test.go:74]: WDYM incorrect? Both the reference hasher and the old bmt hasher give out a different hash for this stuff. What changed in this PR that changes this hash? and will the old hash still be correct?

**Review by acud — APPROVED** [review]

**Review by janos — COMMENTED** [review]
Nice!

I have one suggestion to make comments on exported identifiers comply with https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.

File: pkg/bmt/bmt_test.go [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+		return nil, err
+	}
+	return bmt.Sha3hash(bmt.LengthToSpan(int64(n)), refNoMetaHash)
+}
+
+// Hash hashes the data and the span using the bmt hasher
+func syncHash(h *bmt.Hasher, n int, data []byte) ([]byte, error) {
+	h.Reset()
+	h.SetHeaderInt64(int64(n))
+	_, err := h.Write(data[:n])
```
[the remark this answers is not in the document]
Comment (acud) [reply pkg/bmt/bmt_test.go:57]: Theoretically you're right but when you have a look at how this is used across tests.... this starts to get really ugly. Especially the ones that actually use an uninitialized testData slice :man_shrugging: I tried to do the fix but backed away very quickly since it was getting out of hand. @zelig care to make this pretty? or should we do this at a later PR?

**Review by acud — APPROVED** [review]

**Review by acud — APPROVED** [review]

**Review by janos — APPROVED** [review]
Great! Thanks for addressing my comments.
